### ISSUE TITLE
Add support for .NET Core

### ISSUE THREAD

**Issue by daveaglick** [statiqdev/Statiq#300]
Converted to netstandard2.0/netcoreapp2.0:

**Clients**
* [x] Cake.Wyam

2.x client strategy is discussed in #668

**Core**
* [x] Wyam.Common
* [x] Wyam.Configuration (need to look closely at custom assembly resolution)
* [x] Wyam.Core
* [x] Wyam.Hosting (need to investigate LiveReload support in .NET Standard)
* [x] Wyam.Testing
* [x] Wyam.Testing.JavaScript

**Extensions**
* [x] Wyam.AmazonWebServices
* [x] Wyam.CodeAnalysis
* [x] Wyam.Feeds
* [x] Wyam.GitHub
* [x] Wyam.Highlight
* [x] Wyam.Html
* [x] Wyam.Images
* [x] Wyam.Json
* [x] Wyam.Less
* [x] Wyam.Markdown
* [x] Wyam.Minification
* [x] Wyam.Razor
* [x] Wyam.Sass
* [x] Wyam.SearchIndex
* [x] Wyam.Tables
* [x] Wyam.Xmp
* [x] Wyam.Yaml
* [x] Wyam.YouTube

**Recipes**
* [x] Wyam.Blog
* [x] Wyam.BookSite
* [x] Wyam.Docs
* [x] Wyam.Web

**Client Tests**
* [x] Cake.Wyam.Tests

**Core Tests**
* [x] Wyam.Common.Tests
* [x] Wyam.Configuration.Tests
* [x] Wyam.Core.Tests
* [x] Wyam.Hosting.Tests

**Extension Tests**
* [x] Wyam.CodeAnalysis.Tests
* [x] Wyam.Highlight.Tests
* [x] Wyam.Html.Tests
* [x] Wyam.Json.Tests
* [x] Wyam.Markdown.Tests
* [x] Wyam.Minification.Tests
* [x] Wyam.Razor.Tests
* [x] Wyam.Sass.Tests
* [x] Wyam.Tables.Tests
* [x] Wyam.Xmp.Tests
* [x] Wyam.Yaml.Tests
* [x] Wyam.YouTube.Tests

**Integration Tests**
* [ ] Wyam.Examples.Tests

**Comment by gijswijs** [comment]
Just to be helpful: OctopusDeploy release a tool to help with migrating to Core:

https://icanhasdot.net/result?github=Wyamio~2FWyam

**Comment by daveaglick** [comment]
@gijswijs Thanks, I'll check that out! I'm hoping to start this effort fairly soon. It looks like many of the libraries we depend on have either ported or are in the process.

**Comment by airtonix** [comment]
Yep, i'd love to use this inside docker/linux

**Comment by Vladekk** [comment]
Any progress on this yet?
Docker is on the rise, so Core support would be awesome.

**Comment by pauldotknopf** [comment]
I'd be willing to help out to add .NET Core support.

**Comment by williamwgant** [comment]
Is there a list floating around somewhere of what has to be fixed for Wyam to run on core? Is it possible to generate one somehow?

**Comment by pauldotknopf** [comment]
@daveaglick Can I get started on this now? I need to make wyam as a part of the build process, and I want to use a docker image for the build server :)

**Comment by pauldotknopf** [comment]
I'd probably get the core/tests done first. Then, module-by-module, we will inspect dependencies and determine if we can support .NET Core. I remember seeing a few dependencies that didn't have .NET Core support.

**Comment by daveaglick** [comment]
@pauldotknopf Be my guest! I haven't dug too deeply into it because I suspect some of the assembly context and application domain loading stuff may be a pain until those are introduced in netstandard 2.0. That said, if you're able to get around it, I'd be happy to make some progress here.

**Comment by daveaglick** [comment]
@pauldotknopf Good call - most of the challenging stuff is going to be in `Wyam.Configuration`. If you can get some of the other libraries moved over to netstandard 1.5 now that'll reduce the lead time in any case. And it shouldn't be too disruptive since I just updated everything to Framework 4.6.2 which can consume netstandard 1.5 libraries.

**Comment by philippgille** [comment]
@pauldotknopf @daveaglick Hi, is there any progress regarding Wyam on .NET Core? I didn't try it out, but the [docs](https://wyam.io/docs/usage/obtaining) still say that Wyam runs on .NET Framework 4.6.2

**Comment by daveaglick** [comment]
@philippgille That's correct, it's still full-framework only. Work is progressing on .NET support but it'll still be another couple months. 

We need to convert all the extension libraries to netstandard first, then can convert the core Wyam libraries. netstandard 2.0 is a blocker for some of the core libraries given the way assemblies are dynamically loaded.

Rest assured that .NET Core support is definitely high on the priority list.

**Comment by philippgille** [comment]
The blocking feature is probably `System.AppDomain`? [This document](https://docs.microsoft.com/en-us/dotnet/articles/core/porting/libraries) mentions that `AssemblyLoadContext` is recommended now, although it doesn't say if it's recommended over `AppDomain` in general or only until netstandard 2.0 supports it.

Quote:
> AppDomains can be used for different purposes on the .NET Framework. For code isolation, we recommend separate processes and/or containers as an alternative. For dynamic loading of assemblies, we recommend the new AssemblyLoadContext class.

I didn't work with any of the two namespaces yet, so I don't know if it's a viable alternative for Wyam (or its dependencies).

**Comment by MaximRouiller** [comment]
@daveaglick so.............. ? You waiting for the next major release buddy? 😉 😛 

/poking

**Comment by daveaglick** [comment]
@MaximRouiller Waiting on an enterprising developer to drop a full porting PR in my lap. Hasn't happened yet, but doesn't mean it won't :wink: :wink:

**Comment by MaximRouiller** [comment]
LOLLL

So... apparently, I'm going to be taking vacations soon. How long do you think this is going to take?

**Comment by MaximRouiller** [comment]
Or rather, what are the "steps" that are necessary to do and how many can I do in a weekend?

**Comment by daveaglick** [comment]
To be honest, I'm waiting on netstandard 2.0 to drop. There's a lot going on under the hood with regard to assemblies, dynamic compilation, and binding and 2.0 should have the reflection parts to make that much easier to port. Don't want to spend a ton of time on workarounds now if it'll mostly work with minimal changes soon.

**Comment by MaximRouiller** [comment]
So... it's officially "waiting on 2.0" status.

**Comment by daveaglick** [comment]
Yeah. I just don't have the time to go chasing the bleeding edge.

**Comment by MaximRouiller** [comment]
Yeah, no point in stabbing ourselves in the face with that edge. 😛

**Comment by williamwgant** [comment]
According to this:

https://github.com/dotnet/core/blob/master/roadmap.md

They drop a preview in Q2, with the thing going live in Q3. Assuming less code churn in v2 than in the initial alpha/beta/rc run for core, that means we could be hitting this fairly soon.

**Comment by pauldotknopf** [comment]
Ya, I don't see any reason why this can't be started now.

I have been watching the changes in corefx/coreclr. The work has been really adding back old api's that don't exist for coreclr yet.

**Comment by pauldotknopf** [comment]
Many 3rd party libraries are now supporting .NET Core. The water feels good. Come in.

**Comment by williamwgant** [comment]
What would be the process to do that with this codebase? I've coded in core, but haven't ported anything to it. Do we just pull down the solution and start switching project properties to use netstandard? My impression was that project structures changed significantly going to core and that Microsoft (probably) didn't provide any kind of migration tool. Has that changed?

**Comment by pauldotknopf** [comment]
I have tried the migration stuff in the cli, but it is garbage. It adds a bunch of needless stuff. What you want is a pure/core storing point.

Create a new Wyam.Common project, and copy the csproj into the existing Wyam.Common directory. The code should compile, but with errors. From then, just add references to the packages required, and massage and code that needs to be updated (should be minimal).

Once you have the Wyam.Common project done, you can still reference that csproj file (even though it is .NET Core) from all the existing non-.NET Core projects (add ons, Wyam.Configuration, etc).

It should be an easy and gradual process, with each step being compilable and runnable.

We probably need to identify the order at which the projects should be done.

- Wyam.Common
- Wyam.Core
- TBD

I'd recommend leaving the Wyam.Configuration until later, as you'd probably need the new AppDomain stuff coming. What that means is that initially, the .NET Core version would only support usage directly from the API. No "recipes", etc.

**Comment by pauldotknopf** [comment]
Rough plan for the order of attack.

- Wyam.Common
- Wyam.Core
- Wyam.Hosting
- Any plugin that has dependencies that can be met for .NET Core.

**Comment by pauldotknopf** [comment]
We should probably created a matrix of Wyam extensions and dependencies, and their support for .NET Core. @daveaglick Maybe create a Wiki that we can work on?

**Comment by daveaglick** [comment]
Yeah, I wouldn't touch the automatic migration. It's going to be multi-stage (I've been thinking about the process in preparation for a while). @pauldotknopf has the right idea. Wyam.Common has to go first since it's consumed by everything. The trick is that ported libraries will be able to be used by unported ones but not the other way around.

Here's my thinking in terms of order:
* Wyam.Common (shouldn't be too bad, mostly interfaces and abstractions)
* All module and extension projects that have only netstandard dependencies
* Recipe projects (these can only be done after their dependent module libraries)
* Wyam.Hosting, Wyam.Testing, Wyam.Configuration (that last one may be rough)
* Wyam.Core (do this last since nothing directly consumes it)
* Console app (not sure if I want to multi-target, do a dotnet "tool" alias, or what yet)

A matrix is spot on. I suspect we'll probably end up with at least a couple modules that end up having to stay on full framework, at least initially. That's okay - the whole NuGet/addin system in Wyam should deal with that no problem, but we'll at least want to document it. I don't want to turn the GH wiki on since it might cause confusion with regard to documentation, but I'll add a Markdown file to the root everyone can PR against.

**Comment by daveaglick** [comment]
Also note that as part of this process I'm probably going to remove Wyam.Windows. I'd prefer to get a more integrated NuGet-powered cross-platform upgrading system in place. That's for another issue though.

**Comment by daveaglick** [comment]
https://github.com/Wyamio/Wyam/blob/develop/CorePort.md

**Comment by MaximRouiller** [comment]
Look at me rolling the ball on things I haven't even `git clone` yet. 😜

**Comment by MaximRouiller** [comment]
Thanks @daveaglick for the matrix! 😃

**Comment by daveaglick** [comment]
This may be relevant when it comes time to tackle clients: https://nblumhardt.com/2017/04/netcore-msi/

**Comment by daveaglick** [comment]
It looks like based on https://github.com/aspnet/Home/issues/2022#issuecomment-299548928 that `Wyam.Hosting` may need to target `netcoreapp2.0` because of Kestrel. That'll cascade to consuming apps/libs, so the `Wyam` client will also need to target `netcoreapp2.0`. That probably forces our hand for the kinds of clients we'll be able to build - it's .NET Core or go home (I.e., no multi-targeting to .NET Framework for distribution).

Thankfully it sounds like Razor will at least continue to target `netstandard` so the fallout should be limited to the preview server and client. Once all the ASP.NET Core changes are made we'll need to evaluate any other utility libraries we're bringing in from ASP.NET Core for other extensions and transition away from the ASP.NET Core libraries if they target `netcoreapp2.0`.

I can probably live with the preview server and client targeting `netcoreapp2.0` only, but I would really like to continue targeting `netstandard` for extensions to support the broadest range of embedded scenarios.

**Comment by daveaglick** [comment]
A few updates:
- Got word directly from Scott Hunter yesterday that the intent is for Kestrel to target `netstandard` after all. That still leaves the executable story TBD (should it be a .NET Core app, should we also do a .NET Framework release, etc.) but at least the supporting library story is back to normal.
- There will be some issues with native libraries. Those get placed in a `runtimes` folder in NuGet packages and are typically copied to the app folder based on the RID (see [here](http://www.natemcmaster.com/blog/2016/05/19/nuget3-rid-graph/) and [here](https://docs.microsoft.com/en-us/dotnet/articles/core/rid-catalog)). That all happens at package installation so it's currently unclear how we will support that inside `Wyam.Configuration` for dynamically installed packages. Still thinking about this.

**Comment by daveaglick** [comment]
Follow-up to unmanaged assembly loading - it looks like it can be done by implementing a small special-purpose `AssemblyLoadContext`. Also see [here](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/corehost.md) for lots of information about the process in general.

![2017-05-12_17h58_49](https://cloud.githubusercontent.com/assets/1020407/26018615/f65b14dc-373d-11e7-9ac3-d948725657b0.png)

So if I've got this right, we need to:
- Scan installed NuGet packages (including their dependencies down the heirarchy) for a `runtimes` folder
- If one is found, try to find a valid RID (hopefully some APIs for RID matching)
- When we find the right `runtimes` RID subfolder, load all `.dll` files as unmanaged directly from their paths (no copying needed)

**Comment by Daniel15** [comment]
In the meantime, what's the story with Mono? Does Wyam run in Mono or are there known issues? A lot of .NET Framework apps run fine on Mono, but some apps have issues (particularly if they mess with native code)

> It looks like based on https://github.com/aspnet/Home/issues/2022#issuecomment-299548928 that `Wyam.Hosting` may need to target `netcoreapp2.0` because of Kestrel. That'll cascade to consuming apps/libs, so the `Wyam` client will also need to target `netcoreapp2.0`. That probably forces our hand for the kinds of clients we'll be able to build - it's .NET Core or go home (I.e., no multi-targeting to .NET Framework for distribution).

You can target multiple platforms with one csproj. With [ReactJS.NET](https://reactjs.net/) and [JSPool](https://dan.cx/projects/jspool) I'm targeting both `net40` and `netstandard1.6`:
```
<TargetFrameworks>net40;netstandard1.6</TargetFrameworks>
```

You can just as easily target both `netstandard1.6` and `netcoreapp2.0` if needed. You can use preprocessor directives like `#if` to handle any differences in code - in ReactJS.NET I'm using preprocessor directives to help with sharing code across ASP.NET MVC 5 and ASP.NET Core MVC 1.0.

**Comment by daveaglick** [comment]
@Daniel15 The short story is that Mono doesn't work right now. There's a hard-coded check to exit with an error message if Wyam detects it's being run on Mono since I know it won't work and that reduces support issues.

The long story is that Wyam does some stuff with assembly loading and target framework detection that Mono doesn't seem to support the correctly. I've gotten part of the way towards a resolution but haven't had time to get it fully resolved. I'd like to though - it would be nice to have Mono support in place when I flip the switch to 1.0. I'll probably take one more stab at it.

**Comment by daveaglick** [comment]
@Daniel15 With regard to Kestrel and target frameworks, we thankfully won't need to worry about multi-targeting. The ASP.NET team has walked back their whole "Kestrel is netcoreapp2.0 only" stance and made it clear that it was only netcoreapp for the betas and would target netstandard by the time 2.0 is final. That's great news for Wyam because it means all the supporting libraries except the actual "app(s)" can target netstandard.

**Comment by pauldotknopf** [comment]
I said it before and I will say it again.

For static ~~code~~ **html** generation, this dynamic loading/nuget/appdomain/cake stuff is a bit overkill. Give people a cross-plat API, and let edge projects handle platform glue.

The concept of a static code generation is really simple, yet it currently seems very complicated.

**Comment by daveaglick** [comment]
@pauldotknopf 

> For static code generation, this dynamic loading/nuget/appdomain/cake stuff is a bit overkill

Not sure I follow. Makes sense for cases where you're doing code-gen to include artifacts in some other build process (such as T4), but in this case the code gen is to support a DSL that gets compiled and evaluated. Without support for dynamic loading, NuGet declarations, and the like there would be no way to say "this particular Wyam project needs to use this special package that isn't part of the base distribution" without managing the package dependency by hand. And we'd still have to support adding it as a dynamic reference even then so it could be used from the config file.

Wondering if your thought process is that perhaps there shouldn't even be a Wyam config file and configuring/controlling Wyam execution should/could be handled directly via an API and it's up to the user to determine if they want to use a script, LINQPad, a full project, etc. to call that API?

**Comment by pauldotknopf** [comment]
> Not sure I follow.

Sorry, when I said "code-gen", I meant "html-gen". Converting markdown files to html is a very simple process.

> perhaps there shouldn't even be a Wyam config file and configuring/controlling Wyam execution should/could be handled directly via an API and it's up to the user to determine if they want to use a script, LINQPad, a full project, etc. to call that API?

Exactly. I really need Wyam to run on .NET Core, and it is crazy to me that it can't be done because of this platform glue. The power of Wyam is it's (genius) data model (pipeline/streams) and plugin/extension support. We are needlessly stuck on certain platforms because of glue code. This glue code should exist in *other* repositories/projects, where the weight of the glue doesn't effect the portability of the core API.

**Comment by pauldotknopf** [comment]
Sorry. I know I have hammered this point before, but @Daniel15 triggered me.

**Comment by Silvenga** [comment]
@daveaglick before the move to .NetCore/.NetStandard, Wyam should probably upgrade to the new project SDK - which can be done now (while keeping the .NetFull target). Then the different projects can be slowly migrated to .NetStandard one at a time without impact (should be a drop in replacement in the most cases). Finally the executable projects can move to a multi-target of .NetFull + .NetCore.

.NetStandard 2.0 significantly simplifies this migration with the .NetFull shim.

I've been trying to get a couple of projects upgraded, e.g. https://github.com/raspberry-sharp/raspberry-sharp-io/pull/92

> I propose a three stage migration.
> * Upgrade library projects to new SDK structure - still targeting dotnet 4.0.
> * Move library projects to target dotnet standard 1.6 (or a different level).
> * Move the rest of the projects (tests, etc.) to new SDK structure, keep same targets.

**Comment by MaximRouiller** [comment]
@daveaglick I've seen that done without any major problems on this project: 

https://github.com/terrajobst/nquery-vnext/commits/master 

/ping @terrajobst

**Comment by bbqchickenrobot** [comment]
Any updates?  Is there a branch that is in progress re: the netcore update?  I wouldn't mind helping out =)

**Comment by daveaglick** [comment]
@bbqchickenrobot Making some pretty big progress lately. The last couple weeks leading up to the 1.1.0 release earlier this week have been about switching every project (>40 of them) over to the new SDK MSBuild project system. That paves the way as a first step towards converting to .NET Standard/.NET Core since those targets require the new project format.

Next step is to port Wyam.Common to .NET Standard and then all the other extension libraries can be ported.

**Comment by MaximRouiller** [comment]
@daveaglick I ran out of positive emoji.

Here. Have a pizza. 🍕

**Comment by Silvenga** [comment]
At some point it might be prudent to remove the transitive nuget references in the .csproj's.


For example:

Wyam.Razor needs Microsoft.AspNetCore.Mvc.Razor
Microsoft.AspNetCore.Mvc.Razor needs Microsoft.CodeAnalysis.CSharp
Microsoft.CodeAnalysis.CSharp needs Microsoft.CodeAnalysis.Common

Because the new SDK automatically handles these transitive dependencies automatically we no longer need to reference Microsoft.CodeAnalysis.Common directly.

This will improve future maintainability and prevent bugs in the future.

**Comment by Jetski5822** [comment]
Hey Chaps, was thinking about using the APIs for Orchard Core, but need a net standard version, preferably 2.0 version available. Is there a branch somewhere? Or should I just branch off dev and do PR's if I tackle it?

**Comment by daveaglick** [comment]
@Jetski5822 I've gotten as far as converting the projects to the new SDK format. That should make flipping the switch to target .NET Standard much easier for most of them and I'm hoping to start that work very soon (though help is always appreciated).

I anticipate some challenges with the Wyam.Configuration library given how much it does with assembly loading and reflection, but the others like Wyam.Common and most of the extensions should be ready to port (as long as dependent packages also support .NET Standard). I haven't decided how I want to port the CLI app yet - whether I should target .NET Core, multi-target, make it a .NET Core CLI tool, etc.

I'm curious which APIs you're looking at using in Orchard. I like the project and would be happy to help wherever I can.

**Comment by pauldotknopf** [comment]
@daveaglick, I am starting to come back to this.

I'd like to start migrating the projects. However, I would like to isolate Wyam.Configuration. If users with to  use your "recipe" approach (using Cake), they must use mono or full framework. However, the libraries themselves will reference .NET standard.

Thoughts?

**Comment by daveaglick** [comment]
@pascalberger Woot! I'm totally fine with splitting the work up in stages and doing Wyam.Configuration last. I know that library is going to be tricky for a lot of reasons like the NuGet API stuff and the way assemblies are dynamically resolved. The trick will be to bring all the libraries over to .NET Standard so that the existing path when using configuration files, recipes, etc. continues to work seamlessly but folks who want to start using the main bits on .NET Core can do so.

I'm still wavering a bit on what the next-gen Wyam CLI app should look like. I've very intrigued by the new .NET Core global tools concept (https://blogs.msdn.microsoft.com/dotnet/2018/02/27/announcing-net-core-2-1-preview-1/). My current thoughts are that we'll release both a standard CLI for use as a download or from Chocolaty as well as a .NET Core global tool - hopefully those can be the same project with some different packaging, but that's TBD.

I think you had the right idea in terms of ordering in your comment here: https://github.com/Wyamio/Wyam/issues/300#issuecomment-297500441

**Comment by pascalberger** [comment]
@daveaglick I think you wanted to mention @pauldotknopf

**Comment by daveaglick** [comment]
@pascalberger You are correct, auto-correct fail - though you're welcome to help too :smile:

**Comment by daveaglick** [comment]
After the monster PR from @pauldotknopf I'll continue the conversation here. It looks like we're going to have to be very careful with dependency versions for MS packages right now. Specially, the Razor module is blocking updates. Here's why:

* Razor is totally rewritten for .NET Core 2.0. The API is no where near the same, and it's not even clear if how we're using it will even work, so we're stuck on the current version of Razor libraries for now until I have the time to figure out the new bits.
* That would be fine except that the current version of Razor depends on a large chunk of packages from ASPNET (this is actually one of the things that's supposed to be better in the new version).
* Because Wyam runs all the modules in a single AppDomain, it has to consolidate reference versions across module libraries and only references the latest version from all modules.
* That means if we update packages in _other_ modules, the Razor module will end up binding to versions of it's dependencies that are later than what it's expecting. That would be okay too, except there are some breaking changes between the 1.x versions that Razor references and the 2.x versions.
* Updating to .NET Standard 2.0 looks like it may require updates to some of the same packages that Razor depends on, which as stated above is dangerous and may not even be possible.

All this means that we can't update Razor right now (because it'll be a complicated update that will take time and research) and we can't update packages Razor depends on in _any_ module library (because it'll break Razor).

There's a delicate balance here and I'm still holding out hope that I'll figure out the exact right set of versions to make everything happy. There is a chance though that we'll get totally blocked on updating Razor, at which point I'll start digging in on that problem and we'll have to branch the .NET Standard/Core work until that's done.

**Comment by pauldotknopf** [comment]
Updating core libraries for this many projects/spokes is worse than chaperoning 100 kids to Disney World...

**Comment by MaximRouiller** [comment]
@daveaglick so that is blocking you from upgrading to .NET Core? Anything the .NET team can help you figure out?

**Comment by daveaglick** [comment]
@MaximRouiller Not right now - all the bits are falling into place. .NET Standard 2.0 was a _big_ help. Now that @pauldotknopf has kicked this into high gear it's rolling right along.

* Next release will have most of the extensions, tests, and core libraries converted to .NET Standard.
* Then I've got to take a closer look at my custom assembly resolver and NuGet support and how that'll work in .NET Standard.
* Last bit will be figuring out what the client/exe story should be. Right now I'm leaning towards both a Full Framework exe and a Core exe, a Core global tool, and an old school tool package (for use from Cake).

**Comment by gijswijs** [comment]
I know I'm not much of a help in here, but I just wanted to say that I
value the hard work of everyone involved in getting Wyam to .NET core.
Thank you.



*Gijs van Dam*
*Algemeen Directeur*

*06 52 40 3336*
Mindbus | Full-service internetbureau
3e Binnenvestgracht 23N, 2312 NR Leiden, 071 3015153

**Comment by Vladekk** [comment]
Yep, right now major obstacle in easy adoption of Wyam using cloud pipelines is that free pricing plans and CI platforms usually run on linux. 
So, I myself have to commit output into GIT, not great solution (but works).

Btw, I run huge static blog with Wyam using Gitlab, GIT LFS and Netlify (build locally)
https://gitlab.com/Vladekk/izvne.com

and small static site that is able to be build and deployed using AppVeyor and Netlify 
https://bitbucket.org/Kugelevich/shayana.lv/src

I'd like to help about wyam on .core, but I think I lack deep understanding on .Net required for that.

**Comment by daveaglick** [comment]
@Vladekk Where we're at right now is a little tricky because of the differences in dependencies on .NET Framework and .NET Core (more on that in my next comment). It's at the point where it's probably up to me to push past the remaining issues and that'll open up a ton of opportunities for helping with things like getting Netlify CI set up.

**Comment by daveaglick** [comment]
Status update: Most (though not quite all) of the extension libraries have been ported to .NET Standard. I've started tracking which ones remain in the opening comment of this issue. All the projects have also been converted to the new SDK-style project format and I've removed redundant transitive package references from most (since the new SDK takes care of that).

I'm now working through some nasty dependency issues. The crux of the problem is that we have many different extensions that get loaded at runtime and they have different sets of dependencies, but we can only ever load one version of a given assembly. In .NET Framework we had AppDomains which sort of let you load different versions at once, but those were more trouble than they were worth IMO, .NET Core doesn't have them, and Wyam never used them anyway. Wyam has it's own assembly loading infrastructure to deal with this and it basically worked as first-seen-first-loaded and then any additional requests for an assembly were redirected to whichever version we saw first. In previous versions we mainly lucked out that everything lined up and loading a different version of a dependent assembly than the one a Wyam extension was expecting usually just worked.

Unfortunately after the port things aren't lining up quite as well. For example, Wyam.Less and Wyam.CodeAnalysis depend on Microsoft.Extensions.DependencyInjection 2.0 but Wyam.Razor depends on Microsoft.Extensions.DepenencyInjection 1.1. If version 1.1 gets loaded first then requests for 2.0 get redirected to 1.1 and we end up with `MissingMethodException`. It works the other way too - if 2.0 gets loaded first and there are breaking changes between 1.1 and 2.0 then we have the potential for breaking changes. That said, the problems I'm seeing right now are mostly caused by loading the lower version of a dependent assembly because that's the first one we saw. It looks like I'll have to change the Wyam assembly loading stuff to always load the highest version.

Anyway, we're getting there but it may still be a little while until I'm able to push a new release with the newly .NET Standard extension libraries.

**Comment by Silvenga** [comment]
@daveaglick What are your concerns with LiveReload support under .NetStandard? Maybe I can help.

**Comment by daveaglick** [comment]
@Silvenga Thanks! The biggest problem right now is that Fleck only targets .NET Framework. There's a chance it could still work with .NET Standard/Core 2.0 using the compatibility shim, but I haven't tested that yet (and would prefer not to use the shim if it can be avoided). If the shim doesn't work (and maybe even if it does) we may need to look at using an alternate WebSocket or similar implementation like SignalR.

**Comment by Silvenga** [comment]
@daveaglick I pulled down Fleck and ran the portability analyzer. It appears that Fleck has full API compatibility for .NetStandard 2.0 and .NetCore 2.0. 🎉 I'm so happy for 2.0.

![image](https://user-images.githubusercontent.com/2753330/38042368-d2dab5b6-3279-11e8-85f9-20d1c3a0f6ce.png)

https://docs.microsoft.com/en-us/dotnet/standard/analyzers/portability-analyzer

**Comment by Silvenga** [comment]
@daveaglick created a pull request over on Fleck's side to support .NetStandard - https://github.com/statianzo/Fleck/pull/220.

**Comment by daveaglick** [comment]
@Silvenga Nice! And looks like they’re game to take it in quickly too.

Based on your portability analyzer run, I also did some experiments with switching Wyam.Hosting to netstandard as-is. It looked promising, so may be able to switch even before a Fleck update is published. Just needs a little more testing.

**Comment by escobar5** [comment]
Should Wyam.Razor work on a .NET Core 2.0 app? I get this error when I try to use it:

> TypeLoadException: Could not load type 'Microsoft.AspNetCore.Razor.RazorEngineHost' from assembly 'Microsoft.AspNetCore.Razor, Version=2.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.


As far as I see, RazorEngineHost is not part of Microsoft.AspNetCore.Razor on .Net Core 2:

[Microsoft Docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.razor.razorenginehost?view=aspnetcore-1.1)

**Comment by RLittlesII** [comment]
I am attempting to upgrade `Wyam.Hosting`.  It appears that there is no `netstandard` target for the `owin.dll`.  Has anyone experienced this or am I just missing something.  The following packages won't resolve on a netstandard2.0 target.

- Microsoft.Owin
- Microsoft.Owin.FileSystem
- Microsoft.Owin.StaticFiles
- Owin.Websockets.Fleck

I pointed all of the above to the latest package and it is still telling me that they depend on `owin 1.0`.  Can anyone else confirm, or point me at an issue?  I found this [issue](https://github.com/aspnet/AspNetKatana/issues/27) that discusses the problem.  But I could be off.

**Comment by pauldotknopf** [comment]
Owin is old technology that won't be updated.

Consider it deprecated.

**Comment by RLittlesII** [comment]
@pauldotknopf Then what replaces Owin so I can get Wyam.Hosting ported?

**Comment by daveaglick** [comment]
I’m hopefully circling back around on the porting effort starting next week (I’ve got a Big Thing launching that’s been taking much of my attention recently).

In the meantime, can probably look at https://github.com/natemcmaster/dotnet-serve for an example of how to do this sort of thing under .NET Standard. It basically does exactly what Wyam.Hosting does, just as a global tool. I even ported some of the Wyam.Hosting functionality over to that project, so now we need to port it back to Wyam.Hosting.

The bigger challenge is going to be the Live Reload bits. I _think_ those are .NET Standard so might work out of the box under .NET Core, but we’ll need to be careful there. May end up swapping for SignalR at some point for that.

**Comment by daveaglick** [comment]
@escobar5 Wyam.Razor is a bit of a mess right now. It uses the ASP.NET Core 1.x Razor stuff and the Razor API is totally different in ASP.NET Core 2.x. Not to mention there are a lot of tricky assembly conflicts and breaking changes between the Razor bits currently in Wyam.Razor and the more recent .NET Standard 2.x bits other Wyam module libraries are using.

Presumably the new ASP.NET Core 2.x Razor APIs are simplified but I haven’t had a chance to look closely yet. As part of the port to .NET Standard/Core I want to rewrite the entire Wyam.Razor module to use the new bits before releasing Wyam 2.0 and saying Wyam is .NET Core ready. Right now Wyam is only tested on .NET Framework (though we’re inching closer to .NET Core support).

**Comment by RLittlesII** [comment]
@daveaglick Is your port public?  I'm going to continue hacking at this, and try and help you out a bit.

**Comment by daveaglick** [comment]
Awesome! 👍 yeah, everything in dotnet-serve is oss and my PRs have been merged

**Comment by Daniel15** [comment]
> Then what replaces Owin

@RLittlesII as far as I know, ASP.NET Core is supposed to replace OWIN. You can self host using Kestrel, and ASP.NET Core's middleware API is roughly based off OWIN's

**Comment by MisinformedDNA** [comment]
So what's the deal with using Tabular?

> Wyam.Tables (blocked by Tabular libraries)

The [last version](https://www.nuget.org/packages/Tabular.dll/) came out 7 months ago and before that the last version was in Oct 2014. This project seems mostly dead and, as far as I can tell, is not hosted on GitHub, or anywhere public, and does not have a specified license associated with it.

**Proposal:** Switch to [EPPlus](https://www.nuget.org/packages/EPPlus/) and [CsvHelper](https://www.nuget.org/packages/CsvHelper/).

I have used these for multiple projects and have been very happy with them. Also, Tabular.Excel actually depends on EPPlus, so even they like it.

CsvHelper and EPPlus both are 

- open-source licensed
- available on GitHub
- target .NET Standard 2.0

@daveaglick If you agree, I would like to replace Tabular with EPPlus and CsvHelper.

**Comment by daveaglick** [comment]
@MisinformedDNA That would be awesome - haven't looked too closely at that one yet, so if swapping out Tabular for one of the other libraries is doable, a PR would be great.

**Comment by MisinformedDNA** [comment]
Cool. I'm on it.

**Comment by MisinformedDNA** [comment]
Took a look at Rant and he has [no plans to support .NET Standard](https://github.com/TheBerkin/rant/issues/112). As far as I'm concerned, that's equivalent to saying that the project is dead.

So we either we do the port for him or fork it or we deprecate the `TextGeneration` module. I took a look at it and the tests and I have absolutely no idea what this thing is supposed to be used for. lol

@daveaglick Any thoughts on how you want to handle this?

**Comment by daveaglick** [comment]
@MisinformedDNA Almost certainly deprecate it. It's a fun little module that produces life-like text following user-specified patterns, but AFAIK it never had any real use cases and no one is using it in production. I guess in theory it would have been used for mock-ups and placeholder text, but there are plenty of online generators for that.

**Comment by daveaglick** [comment]
@MisinformedDNA BTW, thanks for jumping in with all the help recently. I've [been off on a little tangent](https://discoverdot.net) but my focus is back here now so I'll hopefully start reviewing PRs and finishing the push towards .NET Core in the next couple days (or next week at the latest depending on how soon I can get some OSS time).

**Comment by eatdrinksleepcode** [comment]
"little tangent" :)

**Comment by daveaglick** [comment]
Getting closer everyone! @MisinformedDNA has helped unblock our remaining modules and @RLittlesII is helping with Wyam.Hosting which is the last remaining core library. The client story is coming together in #668.

I'm going to focus on some bug fixing before getting another release in the 1.x line out the door later this week or next and then that'll probably be it until 2.0 betas (unless other major bugs crop up in 1.x).

**Comment by daveaglick** [comment]
An update (I know it's been a while):

* Wyam.Hosting has been updated to the latest ASP.NET Core 2.x is now fully .NET Standard and no longer uses Fleck for LiveReload support (it's now using raw ASP.NET Core WebSockets)
* Wyam.Razor has been updated to ASP.NET Core 2.x (the Razor engine API totally changed between 1.x and 2.x)

I'm going to do a release in the next couple days (maybe even later today) with the updated Wyam.Hosting and Wyam.Razor bits. At that point _every_ supporting library will be targetting .NET Standard.

The only thing left is the client. I have big plans for how Wyam should work in the future (#668), but I've decided to scale back Wyam 2.0 and make it a straight .NET Core release of what we've already got to get it out the door sooner. I expect the existing client won't take long to port (there isn't that much there) so expect a final Wyam 2.0 release with full .NET Core support in the coming weeks. Then I'll start working on the vNext client story for Wyam 3.0 - and I really hope that doesn't take as long as this did.
